fido/fido.py

```diff
@@ -71,37 +122,32 @@ def connectionLost(self, reason):
             self.finished.errback(reason)
 
 
-def fetch_inner(url, method, headers, body, future, timeout, connect_timeout):
-    """This runs inside a separate thread and orchestrates the async IO
-    work.
+@crochet.run_in_reactor
+def fetch_inner(url, method, headers, body, timeout, connect_timeout):
     """
+    This function must be run in the reactor thread because it is calling
+    twisted API which is not thread safe.
+    See https://crochet.readthedocs.org/en/1.4.0/api.html#run-in-reactor-\
+    asynchronous-results for additional information.
+
+    The implementation of this function with regards to how to use the Twisted
+    Agent is following the official book Twisted Network Programming Essentials
+    chapter 'Web Clients', paragraph 'Agent', page 55 - 2nd Edition.
+
+    :param connect_timeout: maximum time to set up the http connection before
+        aborting. Note that after it is set up, `connect_timeout` loses value.
+        If the server takes forever to send the response and if `timeout` is
+        not provided, we could block forever.
+    :param timeout: maximum time for the server to send a response back before
+        we abort.
+
+    :return: a crochet EventualResult wrapping a
+        twisted.internet.defer.Deferred object
+    """
+
+    from twisted.internet import reactor
```

Comment:
You might want to add a few lines lines here explaining this import; the docstring explains the function and general handling but why are we importing reactor right here and not at the top of the file?